presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java

```diff
@@ -0,0 +1,354 @@
+    {
+        StorageDescriptor storageDescriptor = partition.getSd();
+        if (storageDescriptor == null) {
+            throw new PrestoException(HIVE_INVALID_METADATA, "Partition does not contain a storage descriptor: " + partition);
+        }
+
+        Partition.Builder partitionBuilder = Partition.builder()
+                .setDatabaseName(partition.getDbName())
+                .setTableName(partition.getTableName())
+                .setValues(partition.getValues())
+                .setColumns(storageDescriptor.getCols().stream()
+                        .map(MetastoreUtil::fromMetastoreApiFieldSchema)
+                        .collect(Collectors.toList()))
+                .setParameters(partition.getParameters());
+
+        fromMetastoreApiStorageDescriptor(storageDescriptor, partitionBuilder.getStorageBuilder(), format("%s.%s", partition.getTableName(), partition.getValues()));
+
+        return partitionBuilder.build();
+    }
+
+    private static Properties getHiveSchema(
+            Storage sd,
+            List<Column> dataColumns,
+            List<Column> tableDataColumns,
+            Map<String, String> parameters,
+            String databaseName,
+            String tableName,
+            List<Column> partitionKeys)
+    {
+        // Mimics function in Hive:
+        // MetaStoreUtils.getSchema(StorageDescriptor, StorageDescriptor, Map<String, String>, String, String, List<FieldSchema>)
+
+        Properties schema = new Properties();
+
+        schema.setProperty(hive_metastoreConstants.FILE_INPUT_FORMAT, sd.getStorageFormat().getInputFormat());
+        schema.setProperty(hive_metastoreConstants.FILE_OUTPUT_FORMAT, sd.getStorageFormat().getOutputFormat());
+
+        schema.setProperty(hive_metastoreConstants.META_TABLE_NAME, databaseName + "." + tableName);
+        schema.setProperty(hive_metastoreConstants.META_TABLE_LOCATION, sd.getLocation());
+
+        if (sd.getBucketProperty().isPresent()) {
+            schema.setProperty(hive_metastoreConstants.BUCKET_FIELD_NAME, sd.getBucketProperty().get().getBucketedBy().get(0));
+            schema.setProperty(hive_metastoreConstants.BUCKET_COUNT, Integer.toString(sd.getBucketProperty().get().getBucketCount()));
+        }
+        else {
+            schema.setProperty(hive_metastoreConstants.BUCKET_COUNT, "0");
+        }
+
+        for (Map.Entry<String, String> param : sd.getSerdeParameters().entrySet()) {
+            schema.setProperty(param.getKey(), (param.getValue() != null) ? param.getValue() : "");
+        }
+        schema.setProperty(serdeConstants.SERIALIZATION_LIB, sd.getStorageFormat().getSerDe());
+
+        StringBuilder colNameBuf = new StringBuilder();
+        StringBuilder colTypeBuf = new StringBuilder();
+        StringBuilder colComment = new StringBuilder();
+        boolean first = true;
+        for (Column col : tableDataColumns) {
```

Comment:
for all of these loops try to use `Joiner`. If it doesn't work or is harder to read, that is ok